### PR TITLE
fix smart contract

### PR DESCRIPTION
**Pull request by j-chimienti**
**fixes**:
stake_reward to return _addr not msg.sender
withdraw stake, explicity set to 0.
insert events
init testing

### CODE REVIEW

File: contracts/ColdStaking.sol [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    function start_staking() public payable
+    {
+        assert(msg.value >= staking_threshold);
+        staking_pool = staking_pool.add(msg.value);
+
+        Staker storage _staker = staker[msg.sender];
+
+        _staker.weight = _staker.weight.add(msg.value);
```
Comment (j-chimienti) [inline contracts/ColdStaking.sol]: start_staking wasn't increasing staker weight nor adding value to staking_pool

File: contracts/ColdStaking.sol [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    }
+
+    function withdraw_stake() public only_staker mutex(msg.sender)
+    {
+
+        Staker storage _staker = staker[msg.sender];
+        uint256 _weight = _staker.weight;
+        msg.sender.transfer(_weight);
+        staking_pool = staking_pool.sub(_weight);
+        _staker.weight = 0;
```
Comment (j-chimienti) [inline contracts/ColdStaking.sol]: explicity set staker weight to 0

File: contracts/ColdStaking.sol [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        require(block.number >= staker[msg.sender].init_block.add(round_interval));
+
+        uint256 _reward = stake_reward(msg.sender);
+        msg.sender.transfer(_reward);
+        staker[msg.sender].init_block = block.number;
+
+        Claim(msg.sender, _reward);
+    }
+
+    function stake_reward(address _addr) public view returns (uint256 _reward)
```
Comment (j-chimienti) [inline contracts/ColdStaking.sol]: return stake_reward based on addr, not msg.sender

File: test/ColdStaking.test.js [hunk 4]
```diff
@@ -0,0 +1,107 @@
+// import ether from '../helpers/ether';
```
Comment (j-chimienti) [inline test/ColdStaking.test.js]: needs a lot of work

File: contracts/ColdStaking.sol [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+    event WithdrawStake(address staker, uint256 weight);
+
+    event Claim(address staker, uint256 reward);
+
+    event FirstStakeDonation(address _address, uint256 value);
+
+
+    struct Staker
+    {
+        uint256 weight;
```
Comment (j-chimienti) [inline contracts/ColdStaking.sol]: changing init_block to uint64 would save gas, but worth time / pain?

Comment (Dexaran) [reply contracts/ColdStaking.sol]: Weight is denominated in WEI in fact. As the result, lets keep the value type the same as `msg.value` (uint256)

File: contracts/ColdStaking.sol (hunk 3 above)
Comment (Dexaran) [reply contracts/ColdStaking.sol]: Yes, this makes sense.

File: contracts/ColdStaking.sol (hunk 1 above)
Comment (Dexaran) [reply contracts/ColdStaking.sol]: It seems that it is related to the updated logic of `safeMath.sol` contract.

**Pull request merged**